src/usbhs/mod.rs

```diff
@@ -0,0 +1,434 @@
+            ControlPipe::new(ep0),
+        )
+    }
+}
+
+pub struct Bus<'d, T: Instance> {
+    _phantom: PhantomData<&'d mut T>,
+    fake_power_on: bool,
+}
+
+impl<'d, T: Instance> Bus<'d, T> {
+    fn bus_reset(&mut self) {
+        let regs = T::regs();
+        let d = T::dregs();
+
+        // Reset device address
+        regs.dev_ad().write(|v| {
+            v.set_addr(0);
+        });
+
+        // Reset endpoint state on bus reset
+        // EP0 get ACK/NAK so it can recieve setup
+        d.ep_rx_ctrl(0).write(|v| v.set_mask_uep_r_res(EpRxResponse::ACK));
+        d.ep_tx_ctrl(0).write(|v| v.set_mask_uep_t_res(EpTxResponse::NAK));
+
+        // Mark all other EPs as NAK
+        for i in 1..=15 {
+            d.ep_tx_ctrl(i).write(|v| v.set_mask_uep_t_res(EpTxResponse::NAK));
+            d.ep_rx_ctrl(i).write(|v| v.set_mask_uep_r_res(EpRxResponse::NAK));
+        }
+
+        // Disable all endpoints [1, 15]
+        d.ep_config().write_value(EpConfig::default());
+        d.ep_type().write_value(EpType::default());
+        d.ep_buf_mod().write_value(EpBufMod::default());
+    }
+}
+
+impl<'d, T: Instance> embassy_usb_driver::Bus for Bus<'d, T> {
+    async fn enable(&mut self) {
+        critical_section::with(|_| T::regs().ctrl().modify(|v| v.set_dev_pu_en(true)));
+    }
+
+    async fn disable(&mut self) {}
+
+    async fn poll(&mut self) -> Event {
+        // TODO: VBUS detection
+        if !self.fake_power_on {
+            self.fake_power_on = true;
+            return Event::PowerDetected;
+        }
+        poll_fn(|cx| {
+            BUS_WAKER.register(cx.waker());
+
+            let r = T::regs();
+            let flag = r.int_fg().read();
+
+            if flag.bus_rst() {
+                self.bus_reset();
+                r.int_fg().write(|w| w.set_bus_rst(true));
+                critical_section::with(|_| {
+                    r.int_en().modify(|w| w.set_bus_rst(true));
+                });
+                Poll::Ready(Event::Reset)
+            } else if flag.suspend() {
+                let mis_st = r.mis_st().read();
+                r.int_fg().write(|w| w.set_suspend(true));
+                critical_section::with(|_| {
+                    r.int_en().modify(|w| w.set_suspend(true));
+                });
+                if mis_st.suspend() {
+                    Poll::Ready(Event::Suspend)
+                } else {
+                    Poll::Ready(Event::Resume)
+                }
+            } else {
+                Poll::Pending
+            }
+        })
+        .await
+    }
+
+    fn endpoint_set_enabled(&mut self, ep_addr: EndpointAddress, enabled: bool) {
+        let index = ep_addr.index();
+
+        critical_section::with(|_| match ep_addr.direction() {
+            Direction::Out => {
+                T::dregs().ep_config().modify(|v| v.set_r_en(index - 1, enabled));
+                T::dregs().ep_rx_ctrl(index).write(|v| {
+                    v.set_mask_uep_r_tog(EpTog::DATA1);
```

Comment:
shouldn’t this be DATA0? same for below

Reply:
Ah, I'll switch the data toggling to where NAK is set in the endpoint read/write. This is DATA1 because when ACK is set the data is toggled, so the first one can be DATA0.

Reply:
Yeah we should test with checking the tog_ok, otherwise this is actually not doing much